src/main/java/de/tum/in/www1/artemis/service/metis/PostService.java

```diff
@@ -529,4 +534,35 @@ private void mayInteractWithPostElseThrow(Post post, User user, Course course) {
             authorizationCheckService.checkHasAtLeastRoleInCourseElseThrow(Role.INSTRUCTOR, course, user);
         }
     }
+
+    /**
+     * helper method to retrieve groups and authorities of all posting authors in a list of Posts
+     * @param postsInCourse list of posts with answers whose authors are to be populated with their groups and authorities
+     */
+    private void setAuthorRoleOfPostings(List<Post> postsInCourse) {
+        // sets author roles to display user authority icon on posting headers
+        postsInCourse.forEach(post -> {
+            setAuthorRoleForPosting(post, post.getCoursePostingBelongsTo());
+            post.getAnswers().forEach(answerPost -> setAuthorRoleForPosting(answerPost, post.getCoursePostingBelongsTo()));
+        });
+    }
+
+    /**
+     * fetches and sets groups and authorities of posting author, which are used to display author role icon in the posting header
+     * @param posting       posting to fetch groups and authorities
+     * @param postingCourse course that the post belons to, must be explicitly fetched and provided for new post creation
+     */
+    private void setAuthorRoleForPosting(Posting posting, Course postingCourse) {
```

Comment:
This is not implemented as I suggested and will lead to a bad performance. Imagine there are 50 posts, each with 10 answer posts, then you would execute 500 queries, even if those are all from the same users. In addition, you would load lots of data which is completely unnecessary.

Please instead do the following:
1) collect all unique users (or user ids) for the posts and answer posts you currently deal with (e.g. due after loading one page from the database)
2) use these users (or user ids) and put them into one database query that returns only the user role for those passed user ids. Use the minimal amount of joins to determine this
3) Use the information retrieved from this one single database call and attach it accordingly to all posts and answer posts

Reply:
Thank you. I have implemented your suggestions.